pkg/ipam/ipam.go

```diff
@@ -69,61 +67,68 @@ type Metadata interface {
 }
 
 // NewIPAM returns a new IP address manager
-func NewIPAM(nodeAddressing types.NodeAddressing, c *option.DaemonConfig, owner Owner, localNodeStore *node.LocalNodeStore, k8sEventReg K8sEventRegister, node agentK8s.LocalCiliumNodeResource, mtuConfig MtuConfiguration, clientset client.Clientset) *IPAM {
-	ipam := &IPAM{
+func NewIPAM(nodeAddressing types.NodeAddressing, c *option.DaemonConfig, nodeDiscovery Owner, localNodeStore *node.LocalNodeStore, k8sEventReg K8sEventRegister, node agentK8s.LocalCiliumNodeResource, mtuConfig MtuConfiguration, clientset client.Clientset) *IPAM {
+	return &IPAM{
 		nodeAddressing:   nodeAddressing,
 		config:           c,
 		owner:            map[Pool]map[string]string{},
 		expirationTimers: map[timerKey]expirationTimer{},
 		excludedIPs:      map[string]string{},
+
+		k8sEventReg:    k8sEventReg,
+		localNodeStore: localNodeStore,
+		nodeResource:   node,
+		mtuConfig:      mtuConfig,
+		clientset:      clientset,
+		nodeDiscovery:  nodeDiscovery,
 	}
+}
 
-	switch c.IPAMMode() {
+func (ipam *IPAM) ConfigureAllocator() {
```

Comment:
Can we add some Go doc indicating that this needs to be called before any of the other functions? And maybe add some context when it's safe to call (my understanding is that we mainly want the nodeAddressing to be available)?

To me, this is probably the biggest follow-up item from this work: Figure out how we can untangle the the circular dependencies between Daemon and IPAM, and moving this into a lifecycle start hook.

Since it's now possible to call `Allocate/Release` before the allocators are ready, which would then fail with a nil pointer panic, maybe we should safeguard against that and have those functions return errors?

Reply:
> Can we add some Go doc indicating that this needs to be called before any of the other functions? And maybe add some context when it's safe to call (my understanding is that we mainly want the nodeAddressing to be available)?

Good idea - will add some Go doc to the method.

> 	// Set up ipam conf after init() because we might be running d.conf.KVStoreIPv4Registration

This comment triggered me to keep the call to "ConfigureAllocator" within the daemon initialization. If this is no longer a problem (and might be already solved by dependent components), we could also start introducing the lifecycle aspect and remove this function alltogether. It shouldn't be a big deal to modularize the inner components (allocators and their managers).

> To me, this is probably the biggest follow-up item from this work: Figure out how we can untangle the the circular dependencies between Daemon and IPAM, and moving this into a lifecycle start hook.

Agree, we should move the rest of `daemon/cmd/ipam.go` also into a separate cell as soon as possible. As mentioned in the PR description, it wasn't clear to me where we should put this. As most of the logic is just setting up IPs for health, Ingress etc - and is using IPAM. Definitely something for a follow up.

> Since it's now possible to call Allocate/Release before the allocators are ready, which would then fail with a nil pointer panic, maybe we should safeguard against that and have those functions return errors?

IMO this is already guarded - as the code already checks that the allocator for the corresponding IP family is present. Even thogh the error might be a little bit misleading (`IPv4 allocation disabled` or `IPv6 allocation disabled`)

Reply:
> This comment triggered me to keep the call to "ConfigureAllocator" within the daemon initialization.

I think we can't move it yet. I expect early calls to `ipam.nodeAddressing.IPv4().AllocationCIDR()` to return nil, thus not telling the "host scope" allocator what CIDR to allocate from. But I haven't tested this. 

We receive that information from K8s, so it won't be present after the Daemon lifecycle start hook has started (I believe, would need to double check). But we probably have the start hook wait for `localNodeStore` to announce a non-nil allocation CIDR and initialize the allocator then. Other allocators like the multi-pool allocator already do wait for K8s resources to be synced before they are "ready".

I'm honestly not sure what the `KVStoreIPv4Registration` comment is about. What I can tell you is that we want it to happen after `agentK8s.WaitForNodeInformation` has returned.

> it wasn't clear to me where we should put this.

I'm not sure either, but lifecycle wise we'd want to do it after the allocators are started but before we start handing out IPs to regular pods.

> IMO this is already guarded - as the code already checks that the allocator for the corresponding IP family is present. Even thogh the error might be a little bit misleading (IPv4 allocation disabled or IPv6 allocation disabled)

Ah, perfect!

Reply:
i added a comment above `ConfigureAllocator`